### ISSUE TITLE
x/y-lims are not updated when fitting subplots into layout

### ISSUE THREAD

**Issue by anneaux** [JuliaPlots/Plots.jl#3633]
Hey!
When I fit multiple subplots with equal aspect ratios into a layout, then their axis may be extended in order to fit the layout. However, the values for xlims and ylims are not being updated. Here's a MWE:

```julia
using Plots; gr()
using Random; Random.seed!(1234)

p = plot(layout = grid(1,2), frame=:box)

y1 = rand(10) .* 10
p1 = plot!(sp=1, y1, aspect_ratio=:equal)

y2 = rand(5) .* 10
p2 = plot!(sp=2, y2, aspect_ratio=:equal)
```
This produces this:
![plot1](https://user-images.githubusercontent.com/50271588/124911182-c1b3e400-dfec-11eb-9e0f-449dabaf1516.png)

and I get the following values:
```
extrema(y1) = (2.005860349338411, 8.541465903790503)
Plots.xlims(p1) = (0.73, 10.27)
Plots.ylims(p1) = (1.809792182704848, 8.737534070424065)
extrema(y2) = (0.10905889635595356, 9.567533636029237)
Plots.xlims(p2) = (0.73, 10.27)
Plots.ylims(p2) = (1.809792182704848, 8.737534070424065)
```
You see, for example, that the y axis of subplot 1 (the left one) goes from somewhere -1 to ~12 (ignore the cutted ticks label....), but the according ylims(p1) gives roughly (1.8,8.7). 

So... how can one get the actual limits of the axis after they have been fitted into the layout?

Thanks in advance!

(I'm using Julia v1.6.1, Plots v1.18.0)

**Comment by t-bltg** [comment]
Could you please re-run (for the values and images: I've modified your example to use a fixed seed).

**Comment by t-bltg** [comment]
Stack triggering incorrect values:
```
┌ Debug: 
│   exception =
│    14-element Vector{Base.StackTraces.StackFrame}:
│     macro expansion at logging.jl:341 [inlined]
│     axis_limits(sp::Plots.Subplot{Plots.GRBackend}, letter::Symbol, should_widen::Bool, consider_aspect::Bool) at axes.jl:557
│     axis_limits(sp::Plots.Subplot{Plots.GRBackend}, letter::Symbol) at axes.jl:557
│     get_sp_lims at utils.jl:420 [inlined]
│     ylims(sp::Plots.Subplot{Plots.GRBackend}) at utils.jl:435
│     ylims(plt::Plots.Plot{Plots.GRBackend}, sp_idx::Int64) at utils.jl:445
│     ylims(plt::Plots.Plot{Plots.GRBackend}) at utils.jl:445
│     top-level scope at show.jl:955
│     eval at boot.jl:360 [inlined]
│     include_string(mapexpr::typeof(identity), mod::Module, code::String, filename::String) at loading.jl:1094
│     _include(mapexpr::Function, mod::Module, _path::String) at loading.jl:1148
│     include(mod::Module, _path::String) at Base.jl:386
│     exec_options(opts::Base.JLOptions) at client.jl:285
│     _start() at client.jl:485
└ @ Plots [...]/Plots/CoJwR/src/axes.jl:557
```

**Comment by t-bltg** [comment]
You should use `xlims(p, 2)`, this might seem counter-intuitive but `plot!(sp=2, ...)` does not return the second subplot, but the main plot a.k.a `p`.

```julia
using Plots; gr()
using Random; Random.seed!(1234)

p = plot(layout=grid(1, 2), frame=:box)

y1 = rand(10) .* 10
plot!(sp=1, y1, aspect_ratio=:equal)
@show extrema(y1)
@show xlims(p, 1)
@show ylims(p, 1)

y2 = rand(5) .* 10
plot!(sp=2, y2, aspect_ratio=:equal)
@show extrema(y2)
@show xlims(p, 2)
@show ylims(p, 2)
```
Output:
```
extrema(y1) = (2.005860349338411, 8.541465903790503)
xlims(p, 1) = (0.73, 10.27)
ylims(p, 1) = (1.809792182704848, 8.737534070424065)
extrema(y2) = (0.10905889635595356, 9.567533636029237)
xlims(p, 2) = (0.88, 5.12)
ylims(p, 2) = (-0.17469534583424495, 9.851287878219436)
```

**Comment by anneaux** [comment]
I am sorry to say that, but that does not fix my problem 🙈 you see the values for ylims(p,1)? They still don't match the actual axes limits in the picture!

**Comment by t-bltg** [comment]
> I am sorry to say that, but that does not fix my problem see_no_evil you see the values for ylims(p,1)? They still don't match the actual axes limits in the picture!

I missed that, sorry !

**Comment by t-bltg** [comment]
With `pgfplotsx`:
![eee](https://user-images.githubusercontent.com/13423344/124922915-40634e00-dffa-11eb-8df4-dd2b817cdf1f.png)

So with `gr` is think a margin issue explains with the `1` in `10.0` is chopped out.
But it still doesn't `explain` why the ylims are incorrect.

**Comment by anneaux** [comment]
yes, I don't mind the margins for now ^^ 
It happens for xlims as well (I edited this for clarification in the first post)

**Comment by t-bltg** [comment]
> yes, I don't mind the margins for now ^^
> It happens for xlims as well (I edited this for clarification in the first post)

Ok, so what is not reflected on the plots is:
```
extrema(y1) = (2.005860349338411, 8.541465903790503)
xlims(p, 1) = (0.73, 10.27)
ylims(p, 1) = (1.809792182704848, 8.737534070424065)  # NOK
extrema(y2) = (0.10905889635595356, 9.567533636029237)
xlims(p, 2) = (0.88, 5.12)  # NOK
ylims(p, 2) = (-0.17469534583424495, 9.851287878219436)
```
right ?

**Comment by anneaux** [comment]
exactly!

It seems the axis range in the final plot is determined from either of the two axis and then adjusts the other one accordingly. So that's why always one of them is correctly (x or y)

**Comment by t-bltg** [comment]
The `aspect_ratio` is responsible here, it should be too hard to find the issue:

```julia
using Plots; gr()
using Random; Random.seed!(1234)

p = plot(layout=grid(1, 2), frame=:box, margins=3Plots.mm)

y1 = rand(10) .* 10
plot!(sp=1, y1)

y2 = rand(5) .* 10
plot!(sp=2, y2)
```

gives
![eee](https://user-images.githubusercontent.com/13423344/124926892-5115c300-dffe-11eb-934a-d7713574481b.png)

which is correct.

**Comment by anneaux** [comment]
Ok, I am sorry, maybe I was a little confusing...

I need the plots to keep the aspect ratio, and with the plotting itself I am totally fine!

But, what I want are just the exact values for the axis limits **as they are in the final plot**.
E.g. for relative positioning of annotations as mentioned in issue #2728

**Comment by t-bltg** [comment]
> Ok, I am sorry, maybe I was a little confusing...
> 
> I need the plots to keep the aspect ratio, and with the plotting itself I am totally fine!
> 
> But, what I want are just the exact values for the axis limits **as they are in the final plot**.
> E.g. for relative positioning of annotations as mentioned in issue #2728

This seems to work, triggering rendering of the image before querying the axes limits. This still seems like a bug to me.
```julia
using Plots; gr()
using Random; Random.seed!(1234)

p = plot(layout=grid(1, 2), frame=:box, margins=3Plots.mm)

y1 = rand(10) .* 10
plot!(sp=1, y1, aspect_ratio=:equal)
@show extrema(y1)
@show xlims(p, 1)
@show ylims(p, 1)

y2 = rand(5) .* 10
plot!(sp=2, y2, aspect_ratio=:equal)
@show extrema(y2)
@show xlims(p, 2)
@show ylims(p, 2)

show(devnull, "image/png", p)

@show xlims(p, 1)
@show ylims(p, 1)

@show xlims(p, 2)
@show ylims(p, 2)
```

**Comment by anneaux** [comment]
ohhh thank you! That's indeed a good workaround!

**Comment by t-bltg** [comment]
The root of this bug is getting `plot_ratio = NaN` here https://github.com/JuliaPlots/Plots.jl/blob/4a180989447d7a4e5961ae8b0bdd5bcea02d7f91/src/axes.jl#L601 because of a `[0, 0, 0, 0]` bbox.

**Comment by BeastyBlacksmith** [comment]
> because of a [0, 0, 0, 0] bbox, hence the ylims are not accounting for the :equal aspect ratio.

How did you measure that, `bbox` gives you zeros, if you use it after the fact, but when you add some `@shows` in the line before or above you'd see that they should be non-zero

**Comment by t-bltg** [comment]
> > because of a [0, 0, 0, 0] bbox, hence the ylims are not accounting for the :equal aspect ratio.
> 
> How did you measure that? `bbox` gives you zeros, if you use it after the fact, but when you add some `@shows` in the line before or above you'd see that they should be non-zero

Sorry an all zero `plotarea`.
Adding `@show plot_ratio, plotarea(sp)` after the line quoted before shows:
```
(plot_ratio, plotarea(sp)) = (NaN, BBox{l,t,r,b,w,h = 0.0mm,0.0mm, 0.0mm,0.0mm, 0.0mm,0.0mm})
```

**Comment by t-bltg** [comment]
And adding this
```julia
if isnan(plot_ratio)
    prepare_output(sp.plt)
    plot_ratio = height(plotarea(sp)) / width(plotarea(sp))
end
```

segfaults with a `StackOverflowError` in `_update_min_padding!`.

**Comment by t-bltg** [comment]
We could also change `get_sp_lims` in `utils.jl` to:
```julia
function get_sp_lims(sp::Subplot, letter::Symbol)
    if height(plotarea(sp)) == 0 || width(plotarea(sp)) == 0
        prepare_output(sp.plt)
    end
    axis_limits(sp, letter)
end
```

**Comment by BeastyBlacksmith** [comment]
I don't think that is the right way to fix this. There is some calculation in `prepare_output` that should move upwards in the pipeline such that `xlims(sp)` gives the correct limits regardless of the `aspect_ratio`

**Comment by t-bltg** [comment]
> I don't think that is the right way to fix this. There is some calculation in `prepare_output` that should move upwards in the pipeline such that `xlims(sp)` gives the correct limits regardless of the `aspect_ratio`

The problem is `_update_min_padding!` which is backend dependent.

**Comment by BeastyBlacksmith** [comment]
why is it a problem that it is backend dependent?

**Comment by t-bltg** [comment]
> why is it a problem that it is backend dependent?

Because then you cannot fix https://github.com/JuliaPlots/Plots.jl/issues/3633#issuecomment-876503495 which is not backend dependent (no show, savefig, display, ...)

**Comment by BeastyBlacksmith** [comment]
There are quite a few functions in the pipeline before show that are backend dependent (all the ones in `templates/backend.jl`) though, not every backend defines all of them

**Comment by BeastyBlacksmith** [comment]
I agree that you should get the right answer independent of the backend, but in general it would be fine to get a different answer for different backends (probably not in this particular case)

**Comment by t-bltg** [comment]
> I agree that you should get the right answer independent of the backend, but in general it would be fine to get a different answer for different backends (probably not in this particular case)

Yes it is normal to get a different answer for a different backend. However I believe that the issue exposed in the mwe https://github.com/JuliaPlots/Plots.jl/issues/3633#issuecomment-876503495 should be fixed, we can't leave it like this.
